storage/innobase/log/log0log.cc

```diff
@@ -778,31 +777,29 @@ inline void log_t::persist(lsn_t lsn) noexcept
 }
 #endif
 
+ATTRIBUTE_COLD ATTRIBUTE_NOINLINE
 /** Write resize_buf to resize_log.
 @param length  the used length of resize_buf */
-ATTRIBUTE_COLD void log_t::resize_write_buf(size_t length) noexcept
+void log_t::resize_write_buf(size_t length) noexcept
 {
-  const size_t block_size_1= get_block_size() - 1;
+  const size_t block_size_1= write_size - 1;
   ut_ad(!(resize_target & block_size_1));
   ut_ad(!(length & block_size_1));
   ut_ad(length > block_size_1);
   ut_ad(length <= resize_target);
   const lsn_t resizing{resize_in_progress()};
   ut_ad(resizing <= write_lsn);
-  lsn_t offset= START_OFFSET +
-    ((write_lsn - resizing) & ~lsn_t{block_size_1}) %
-    (resize_target - START_OFFSET);
+  lsn_t offset= START_OFFSET;
 
   if (UNIV_UNLIKELY(offset + length > resize_target))
-  {
-    offset= START_OFFSET;
     resize_lsn.store(first_lsn +
                      (~lsn_t{block_size_1} & (write_lsn - first_lsn)),
                      std::memory_order_relaxed);
-  }
+  else
+    offset+= (write_lsn - resizing) & ~lsn_t{block_size_1};
```

Comment:
You are right.  There are two problems here. First, we have to assign the correct `offset` before we can check it.

If the condition holds and we would be about to wrap around the `ib_logfile101`, we could simply wrap straight to `START_OFFSET` (12288), but then we would have to adjust the `resize_lsn` to the straight `write_lsn`, instead of retaining the modulo arithmetic logic that we had around the original initialization of `offset`, like this:
```diff
-  lsn_t offset= START_OFFSET +
-    ((write_lsn - resizing) & ~lsn_t{block_size_1}) %
-    (resize_target - START_OFFSET);
+  lsn_t offset= START_OFFSET + (write_lsn - resizing) & ~lsn_t{block_size_1};
 
   if (UNIV_UNLIKELY(offset + length > resize_target))
   {
     offset= START_OFFSET;
-    resize_lsn.store(first_lsn +
-                     (~lsn_t{block_size_1} & (write_lsn - first_lsn)),
-                     std::memory_order_relaxed);
+    resize_lsn.store(write_lsn, std::memory_order_relaxed);
   }
```
This perfectly explains some resizing related trouble in #3327. Why is correct to remove the `% (resize_target - START_OFFSET)` computation? Normally, the left-hand-side of the modulus would be less than `resize_target - START_OFFSET` and therefore the operation has no effect. If the left-hand-side is greater, then also `offset + length > resize_target` will hold and we will perform a simplified wrap-around logic in the `if` block.

The above revised logic corresponds to the following in `log_t::resize_start()` where we will assign a `start_lsn` that will correspond to the `START_OFFSET`:
```c++
        else
        {
          memcpy_aligned<16>(resize_buf, buf, (buf_free + 15) & ~15);
          start_lsn= first_lsn +
            (~lsn_t{write_size - 1} & (write_lsn - first_lsn));
        }
      }
      resize_lsn.store(start_lsn, std::memory_order_relaxed);
```
The computation of `start_lsn` refers to the `log_sys.write_lsn` of the current `ib_logfile0` when the resizing (the creation of `ib_logfile101`) starts.

Anyway, because this is not time critical code, I agree that we’d better avoid this change now so that we will not depend on any additional stress testing to validate this change. We have the regression test `innodb.log_file_size_online` which does restart the server, but the log file would typically not wrap around during the execution of that test.